prism_compile.c

```diff
@@ -1195,60 +1242,706 @@ pm_compile_multi_write_lhs(rb_iseq_t *iseq, NODE dummy_line_node, const pm_node_
+
+    ADD_INSN(ret, &line.node, pop);
+    ADD_INSN(ret, &line.node, pop);
+    ADD_LABEL(ret, match_succeeded_label);
+
+    return COMPILE_OK;
+}
+
+/**
+ * This function generates the code to set up the error string and error_p
+ * locals depending on whether or not the pattern matched when the value needs
+ * to match a specific deconstructed length.
+ */
+static int
+pm_compile_pattern_length_error(rb_iseq_t *iseq, pm_scope_node_t *scope_node, const pm_node_t *node, LINK_ANCHOR *const ret, VALUE message, VALUE length, unsigned int base_index)
+{
+    pm_line_node_t line;
+    pm_line_node(&line, scope_node, node);
+
+    LABEL *match_succeeded_label = NEW_LABEL(line.lineno);
+
+    ADD_INSN(ret, &line.node, dup);
+    ADD_INSNL(ret, &line.node, branchif, match_succeeded_label);
+
+    ADD_INSN1(ret, &line.node, putspecialobject, INT2FIX(VM_SPECIAL_OBJECT_VMCORE));
+    ADD_INSN1(ret, &line.node, putobject, message);
+    ADD_INSN1(ret, &line.node, topn, INT2FIX(3));
+    ADD_INSN(ret, &line.node, dup);
+    ADD_SEND(ret, &line.node, idLength, INT2FIX(0));
+    ADD_INSN1(ret, &line.node, putobject, length);
+    ADD_SEND(ret, &line.node, id_core_sprintf, INT2FIX(4));
+    ADD_INSN1(ret, &line.node, setn, INT2FIX(base_index + PM_PATTERN_BASE_INDEX_OFFSET_ERROR_STRING + 1));
+
+    ADD_INSN1(ret, &line.node, putobject, Qfalse);
+    ADD_INSN1(ret, &line.node, setn, INT2FIX(base_index + PM_PATTERN_BASE_INDEX_OFFSET_KEY_ERROR_P + 2));
+
+    ADD_INSN(ret, &line.node, pop);
+    ADD_INSN(ret, &line.node, pop);
+    ADD_LABEL(ret, match_succeeded_label);
+
+    return COMPILE_OK;
+}
+
+/**
+ * This function generates the code to set up the error string and error_p
+ * locals depending on whether or not the pattern matched when the value needs
+ * to pass a specific #=== method call.
+ */
+static int
+pm_compile_pattern_eqq_error(rb_iseq_t *iseq, pm_scope_node_t *scope_node, const pm_node_t *node, LINK_ANCHOR *const ret, unsigned int base_index)
+{
+    pm_line_node_t line;
+    pm_line_node(&line, scope_node, node);
+
+    LABEL *match_succeeded_label = NEW_LABEL(line.lineno);
+
+    ADD_INSN(ret, &line.node, dup);
+    ADD_INSNL(ret, &line.node, branchif, match_succeeded_label);
+
+    ADD_INSN1(ret, &line.node, putspecialobject, INT2FIX(VM_SPECIAL_OBJECT_VMCORE));
+    ADD_INSN1(ret, &line.node, putobject, rb_fstring_lit("%p === %p does not return true"));
+    ADD_INSN1(ret, &line.node, topn, INT2FIX(3));
+    ADD_INSN1(ret, &line.node, topn, INT2FIX(5));
+    ADD_SEND(ret, &line.node, id_core_sprintf, INT2FIX(3));
+    ADD_INSN1(ret, &line.node, setn, INT2FIX(base_index + PM_PATTERN_BASE_INDEX_OFFSET_ERROR_STRING + 1));
+    ADD_INSN1(ret, &line.node, putobject, Qfalse);
+    ADD_INSN1(ret, &line.node, setn, INT2FIX(base_index + PM_PATTERN_BASE_INDEX_OFFSET_KEY_ERROR_P + 2));
+    ADD_INSN(ret, &line.node, pop);
+    ADD_INSN(ret, &line.node, pop);
+
+    ADD_LABEL(ret, match_succeeded_label);
```

Comment:
Why is the `match_succeeded_label` in a different place here than the above methods?

Reply:
I believe because in this case it checks the result of matching the pattern, and if it fails then it sets `error_p` to false.